### PR TITLE
Room build out

### PR DESCRIPTION
**Pull request by ryanquinn3**
Okay so I got clicking on a library song to add to the song to the current rooms queue. 

Reference PrivateQueue.ts for below:

We need a way to automatically load the first song added to a queue into the player. This can be done a couple of ways. Want your thoughts. Initially I thought about using an observable on the queue class itself. I started implementing it in PrivateQueue (its all commented out). The more I thought about it I feel like we shouldn't put that functionality into Queue but the queue component and have it communicate to the roomComponent via an event. (See PlayerComponent with the @Ouput decorator / syntax). Was thinking something along the lines of firstSongAdded (event) that fires when the queue was empty but now is not. Well still need to add something to queue class to communicate that something has changed with its list but we could just do a callback over an observable. Lemme know what you think.

Also was thinking about it and lets stick to the following convention for export / import.

If exporting an angular service/component use default export. If you're exporting a generic class / type / interface use named exports( { YourClass } syntax). We want class/type names to be consistent throughout our codebase.

### CODE REVIEW

**Comment by dswen** [comment]
I'm down with the export/import style you mention.

File: src/public/app/room/room.component.ts [hunk 1]
```diff
@@ -1,9 +1,13 @@
-import { Component } from 'angular2/core';
+import {Component, OnChanges, SimpleChange} from 'angular2/core';
```
Comment (ryanquinn3) [inline src/public/app/room/room.component.ts]: This isnt needed, ill remove

File: src/public/app/core/types/ScPlayer.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    Pause = 'pause',
+    Finished = 'finish',
+    Seek = 'seek',
+    Seeked = 'seeked',
+    GeoBlocked = 'geo_blocked',
+    BufferingStart = 'buffering_start',
+    BufferingEnd = 'buffering_end',
+    AudioError = 'audio_error',
+    Time = 'time',
+    NoStreams = 'no_streams',
```
Comment (dswen) [inline src/public/app/core/types/ScPlayer.ts:25]: let's do BLOCK_CAPS for enums like in java.

File: src/public/app/core/types/ScPlayer.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+export interface ScPlayer {
```
Comment (dswen) [inline src/public/app/core/types/ScPlayer.ts:1]: Is this specific to Soundcloud? Should it just be called Player?

File: src/public/app/library/library.service.ts [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
-export default class LibraryService {
-      constructor(public http: Http) {}
+class LibraryService {
+    constructor(public http: Http) {}
 
-      public getLibrary(): Observable<any> {
-        // noinspection TypeScriptUnresolvedFunction
-            return this.http.get('/api/').map((songs: any) => songs.json());
-      }
+    public getLibrary(): Observable<any> {
```
Comment (dswen) [inline src/public/app/library/library.service.ts]: Can you get rid of the "any"s here? What type of Observable does this return? Also what's the type of songs that comes back?

File: src/public/app/player/player.component.ts [hunk 5]
```diff
@@ -25,20 +23,23 @@ const template: string = `
     template,
     selector: 'player',
 })
-export class Player implements OnChanges {
-
-    @Input() private song: Song;
+class Player implements OnChanges {
```
Comment (dswen) [inline src/public/app/player/player.component.ts:26]: What's the difference between this and ScPlayer?

File: src/public/app/core/types/ScPlayer.ts (hunk 3 above)
Comment (ryanquinn3) [inline src/public/app/core/types/ScPlayer.ts:1]: Yes. We have their sdk which has a method stream. Stream returns a promise with this player instance. This player is what actually plays the music. I figured prefixing all soundcloud provided things with `Sc`

File: src/public/app/player/player.component.ts [hunk 6]
```diff
@@ -25,20 +23,23 @@ const template: string = `
...
 
+    @Input() public queuedSong: QueuedUpSong;
+    @Output() public songFinished: EventEmitter = new EventEmitter();
     private player: ScPlayer;
```
Comment (dswen) [inline src/public/app/player/player.component.ts:30]: I think instead of putting an ScPlayer on the Player object, we should have the Song interface have a getPlayer function. That way we don't have to do

if (song.isSoundcloud) {
  scPlayer.play(song);
} else if (song.isSpotify) {
  spotifyPlayer.play(song);
}

File: src/public/app/player/player.component.ts (hunk 5 above)
Comment (dswen) [inline src/public/app/player/player.component.ts:26]: never mind

File: src/public/app/room/PrivateQueue.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        if (!this.songs.length) {
+            return null;
+        }
+        let nextSong: QueuedUpSong = this.songs.shift();
+        // this.songsObserver.next(this.songs);
+        return nextSong;
+    }
+
+    public removeSongAtIndex(index: number): void {
+        this.songs = this.songs.filter( (song: QueuedUpSong, i: number) => {
```
Comment (dswen) [inline src/public/app/room/PrivateQueue.ts:37]: I think you're abusing filter here. Why not just this.songs.splice(index, 1) ?

Comment (ryanquinn3) [inline src/public/app/room/PrivateQueue.ts:37]: I did this because angular2 recommends immutable data structures. It helps with change detection to have to store only the pointer to the main object rather than having to figure out if something within the data structure changed.

http://victorsavkin.com/post/110170125256/change-detection-in-angular-2

File: src/public/app/room/room.service.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+@Injectable()
+class RoomService {
+
+    private currentRoom: Room = null;
+    private allRooms: Room[] = [];
+
+    constructor(private http: Http){
+        //want to ping server for rooms, just testing with private room now
+        this.allRooms.push(new PrivateRoom());
```
Comment (dswen) [inline src/public/app/room/room.service.ts]: nit: I think it's clearer to have

this.allRooms = [new PrivateRoom()]; 

instead of declaring above and adding to it in the constructor.

File: src/public/app/player/player.component.ts (hunk 6 above)
Comment (ryanquinn3) [inline src/public/app/player/player.component.ts:30]: I'm not against that. We can probably use that to pre-fetch the players too and that way our playerComponent will just call `.play()`. Lets do that in another PR though.

File: src/public/app/room/PrivateQueue.ts (hunk 7 above)
Comment (dswen) [inline src/public/app/room/PrivateQueue.ts:37]: Word. Immutable things are good.
